core/src/main/java/cz/o2/proxima/scheme/AttributeValueAccessors.java

```diff
@@ -0,0 +1,148 @@
+
+    /**
+     * Create value from object
+     *
+     * @param object create from
+     * @return value
+     */
+    default T createFrom(Object object) {
+      throw new UnsupportedOperationException("Method createFrom() is not implemented.");
+    }
+
+    /**
+     * Get value @TODO: this is a little bit weird
+     *
+     * @param value value
+     * @param <V> expected type
+     * @return value
+     */
+    @SuppressWarnings("unchecked")
+    default <V> V valueOf(T value) {
+      return (V) value;
+    }
+  }
+
+  public interface PrimitiveValueAccessor<T> extends GenericValueAccessor<T> {}
+
+  /**
+   * Array value accessor
+   *
+   * @param <T>
+   */
+  public interface ArrayValueAccessor<T> extends GenericValueAccessor<T> {
+
+    default <V> T[] createFrom(V[] object) {
+      throw new UnsupportedOperationException("Method createFrom() is not implemented.");
+    }
+
+    default <V> V[] valuesOf(T[] object) {
+      throw new UnsupportedOperationException("Method valuesOf() is not implemented.");
+    }
+
+    default <V> V[] valuesOf(T object) {
+      throw new UnsupportedOperationException("Method valuesOf() is not implemented.");
+    }
+  }
+
+  public static class DefaultArrayValueAccessor<T> implements ArrayValueAccessor<T> {
+
+    private final GenericValueAccessor<T> valueAccessor;
+
+    public DefaultArrayValueAccessor(GenericValueAccessor<T> valueAccessor) {
+      this.valueAccessor = valueAccessor;
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public <V> T[] createFrom(V[] object) {
+      return (T[]) Arrays.stream(object).map(valueAccessor::createFrom).toArray();
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public <V> V[] valuesOf(T[] object) {
+      return (V[]) Arrays.stream(object).map(valueAccessor::valueOf).toArray();
+    }
+  }
+
+  public interface StructureValueAccessor<T> extends GenericValueAccessor<T> {
+
+    Map<String, Object> valuesOf(T value);
+
+    <V> V readField(String name, T value);
+
+    T createFrom(Map<String, Object> map);
+  }
+
+  public interface EnumValueAccessor<T> extends GenericValueAccessor<T> {}
+
+  public static class DefaultStructureValueAccessor<T> implements StructureValueAccessor<T> {
+
+    @SuppressWarnings("unchecked")
+    public Map<String, Object> valuesOf(T value) {
+      checkInputValue(value);
+      return (Map<String, Object>) value;
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public <V> V readField(String name, T value) {
+      checkInputValue(value);
+      return ((Map<String, V>) value).get(name);
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public T createFrom(Map<String, Object> map) {
+      return (T) map;
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public T createFrom(Object object) {
+      checkInputValue(object);
+      return createFrom((Map<String, Object>) object);
+    }
+
+    private void checkInputValue(Object object) {
```

Comment:
Why can't we have `Map<String, Object>` in a parameter signature instead?

Reply:
This is because of `Array<Struct>` but I'll try to find better solution.

Reply:
Related to the suggested simplication.